File: benchmark_python/progress.py
"""Progress logger: periodically logs insert and query counts while run is active (async)."""
from __future__ import annotations

import asyncio
import logging
from typing import Any

logger = logging.getLogger(__name__)

RESET = "\033[0m"
DIM = "\033[2m"
CYAN = "\033[36m"
YELLOW = "\033[33m"
GREEN = "\033[32m"
WHITE = "\033[37m"

COL_W = 12
PREFIX = "           "  # 11 chars to align with "  Insert   " / "  Query    "


def _pad_left(s: str, w: int) -> str:
    """Right-align s in column of width w (pad with spaces on the left)."""
    if len(s) >= w:
        return s
    return s.rjust(w)


def _fmt_insert_header() -> str:
    """Insert row: status (incoming, completed) | interval (int_) | cumulative (cum_)."""
    return (
        f"{YELLOW}  Insert   "
        f"{_pad_left('incoming', COL_W)}{_pad_left('completed', COL_W)} "
        f"{_pad_left('int_tot', COL_W)}{_pad_left('int_orig', COL_W)}"
        f"{_pad_left('int_dup', COL_W)}{_pad_left('int_avg_ms', COL_W)} "
        f"{_pad_left('cum_tot', COL_W)}{_pad_left('cum_orig', COL_W)}"
        f"{_pad_left('cum_dup', COL_W)}{_pad_left('cum_avg_ms', COL_W)}"
        f"{RESET}"
    )


def _fmt_insert_data(
    incoming: int,
    completed: int,
    interval_total: int,
    interval_originals: int,
    interval_duplicates: int,
    interval_avg_insert_ms: float,
    total: int,
    originals: int,
    duplicates: int,
    cumulative_avg_insert_ms: float,
) -> str:
    """One data row for Insert: incoming, completed, then interval and cumulative (same as Go)."""
    return (
        f"{PREFIX}"
        f"{CYAN}{incoming:>{COL_W}d}{RESET}{CYAN}{completed:>{COL_W}d}{RESET} "
        f"{CYAN}{interval_total:>{COL_W}d}{RESET}{CYAN}{interval_originals:>{COL_W}d}{RESET}"
        f"{CYAN}{interval_duplicates:>{COL_W}d}{RESET}{CYAN}{interval_avg_insert_ms:>{COL_W}.2f}{RESET} "
        f"{CYAN}{total:>{COL_W}d}{RESET}{CYAN}{originals:>{COL_W}d}{RESET}"
        f"{CYAN}{duplicates:>{COL_W}d}{RESET}{CYAN}{cumulative_avg_insert_ms:>{COL_W}.2f}{RESET}"
    )


def _fmt_query_header() -> str:
    """Query row: interval (int_) | cumulative (cum_)."""
    return (
        f"{YELLOW}  Query    "
        f"{_pad_left('int_queries', COL_W)}{_pad_left('int_failed', COL_W)}{_pad_left('int_avg_ms', COL_W)} "
        f"{_pad_left('cum_queries', COL_W)}{_pad_left('cum_failed', COL_W)}{_pad_left('cum_avg_ms', COL_W)}"
        f"{RESET}"
    )


def _fmt_query_data(
    interval_q: int,
    interval_failed: int,
    interval_avg_ms: float,
    q: int,
    failed: int,
    avg_latency_ms: float,
) -> str:
    """One data row for Query: 6 right-aligned values, cyan."""
    return (
        f"{PREFIX}"
        f"{CYAN}{interval_q:>{COL_W}d}{RESET}{CYAN}{interval_failed:>{COL_W}d}{RESET}"
        f"{CYAN}{interval_avg_ms:>{COL_W}.2f}{RESET} "
        f"{CYAN}{q:>{COL_W}.0f}{RESET}{CYAN}{failed:>{COL_W}.0f}{RESET}"
        f"{CYAN}{avg_latency_ms:>{COL_W}.2f}{RESET}"
    )

async def run_progress_logger(
    inserted_lock: asyncio.Lock,
    inserted_shared: list[float],
    stop_event: asyncio.Event,
    queries_lock: asyncio.Lock,
    queries_shared: list[float],
    interval_sec: float = 5.0,
) -> None:
    """Log insert/query counts every interval_sec until stop_event is set (same columns as Go)."""
    prev_inserted = [0.0, 0.0, 0.0, 0.0, 0.0]
    prev_insert_started = 0.0
    prev_queries = 0.0
    prev_query_latency_sec = 0.0
    prev_failed = 0.0
    while not stop_event.is_set():
        try:
            await asyncio.wait_for(stop_event.wait(), timeout=interval_sec)
        except asyncio.TimeoutError:
            pass
        if stop_event.is_set():
            break
        async with inserted_lock:
            total = inserted_shared[0]
            originals = inserted_shared[1]
            duplicates = inserted_shared[2]
            total_insert_latency_sec = inserted_shared[3]
            insert_statements = inserted_shared[4]
            cur_insert_started = inserted_shared[6]
        async with queries_lock:
            q = int(queries_shared[0])
            total_latency_sec = queries_shared[1]
            failed = queries_shared[2]
        interval_insert_started = int(cur_insert_started - prev_insert_started)
        prev_insert_started = cur_insert_started
        interval_total = int(total - prev_inserted[0])
        interval_originals = int(originals - prev_inserted[1])
        interval_duplicates = int(duplicates - prev_inserted[2])
        interval_latency_sec = total_insert_latency_sec - prev_inserted[3]
        interval_statements = int(insert_statements - prev_inserted[4])
        prev_inserted[:] = [total, originals, duplicates, total_insert_latency_sec, insert_statements]
        interval_avg_insert_ms = (interval_latency_sec / interval_total * 1000.0) if interval_total > 0 else 0.0
        cumulative_avg_insert_ms = (total_insert_latency_sec / total * 1000.0) if total > 0 else 0.0
        interval_q = q - int(prev_queries)
        interval_query_latency_sec = total_latency_sec - prev_query_latency_sec
        interval_failed = int(failed - prev_failed)
        prev_queries = q
        prev_query_latency_sec = total_latency_sec
        prev_failed = failed
        avg_latency_ms = (total_latency_sec / q * 1000.0) if q > 0 else 0.0
        interval_avg_ms = (interval_query_latency_sec / interval_q * 1000.0) if interval_q > 0 else 0.0
        logger.info("%s---%s", DIM, RESET)
        logger.info("%s", _fmt_insert_header())
        logger.info(
            "%s",
            _fmt_insert_data(
                interval_insert_started,
                interval_statements,
                interval_total,
                interval_originals,
                interval_duplicates,
                interval_avg_insert_ms,
                int(total),
                int(originals),
                int(duplicates),
                cumulative_avg_insert_ms,
            ),
        )
        logger.info("%s", _fmt_query_header())
        logger.info(
            "%s",
            _fmt_query_data(interval_q, interval_failed, interval_avg_ms, q, int(failed), avg_latency_ms),
        )

File: benchmark-go/progress.go
package benchmarkgo

import (
	"fmt"
	"log"
	"sync/atomic"
	"time"
)

const defaultInterval = 5 * time.Second

// Atomic counters (int64). Latencies stored in microseconds for atomic Add.
var (
	insertTotal         atomic.Int64
	insertOriginals     atomic.Int64
	insertDuplicates    atomic.Int64
	insertLatencyMicros atomic.Int64
	insertStatements    atomic.Int64
	insertStarted       atomic.Int64
	insertPostgres1     atomic.Int64 // rows inserted via pgbouncer.database=postgres1
	insertPostgres2     atomic.Int64 // rows inserted via pgbouncer.database=postgres2
	queryCount          atomic.Int64
	queryLatencyMicros  atomic.Int64
	queryFailed         atomic.Int64
)

func init() {
	// No need to zero - Go zero-inits atomics
}

// AddInsert records an insert batch. Latency is in microseconds.
func AddInsert(total, originals, duplicates, latencyMicros, statements int64) {
	insertTotal.Add(total)
	insertOriginals.Add(originals)
	insertDuplicates.Add(duplicates)
	insertLatencyMicros.Add(latencyMicros)
	insertStatements.Add(statements)
}

// AddInsertStarted records one batch handed to a worker (incoming).
func AddInsertStarted(delta int64) {
	insertStarted.Add(delta)
}

// AddInsertToDB records rows inserted for a specific PgBouncer database (postgres1 or postgres2). No-op if db is empty.
func AddInsertToDB(db string, count int64) {
	switch db {
	case "postgres1":
		insertPostgres1.Add(count)
	case "postgres2":
		insertPostgres2.Add(count)
	}
}

// AddQuery records a query batch. Latency is in microseconds.
func AddQuery(count, latencyMicros, failed int64) {
	queryCount.Add(count)
	queryLatencyMicros.Add(latencyMicros)
	queryFailed.Add(failed)
}

// padRight returns s padded with spaces on the right to width w.
func padRight(s string, w int) string {
	if len(s) >= w {
		return s
	}
	return s + fmt.Sprintf("%*s", w-len(s), "")
}

// padLeft returns s padded with spaces on the left to width w (right-aligned in column).
func padLeft(s string, w int) string {
	if len(s) >= w {
		return s
	}
	return fmt.Sprintf("%*s", w, s)
}

// ANSI colors for progress output (omit if not a TTY for pipe-friendly logs)
const (
	_colorReset  = "\033[0m"
	_colorDim    = "\033[2m"
	_colorCyan   = "\033[36m"
	_colorYellow = "\033[33m"
	_colorGreen  = "\033[32m"
)

// Snapshot is the final aggregated state, sent on resultCh when doneCh is closed.
type Snapshot struct {
	Inserted InsertedStats
	Queries  QueryStats
}

// InsertedStats holds aggregated insert stats.
type InsertedStats struct {
	Total                 float64
	Originals             float64
	Duplicates            float64
	TotalInsertLatencySec float64
	InsertStatements      float64
	Postgres1             float64 // rows inserted via pgbouncer.database=postgres1
	Postgres2             float64 // rows inserted via pgbouncer.database=postgres2
}

// QueryStats holds aggregated query stats.
type QueryStats struct {
	Count           float64
	TotalLatencySec float64
	FailedCount     float64
}

// loadSnapshot reads current atomic counters into a Snapshot (latency from micros to sec).
func loadSnapshot() Snapshot {
	insLat := insertLatencyMicros.Load()
	qLat := queryLatencyMicros.Load()
	return Snapshot{
		Inserted: InsertedStats{
			Total:                 float64(insertTotal.Load()),
			Originals:             float64(insertOriginals.Load()),
			Duplicates:             float64(insertDuplicates.Load()),
			TotalInsertLatencySec: float64(insLat) / 1e6,
			InsertStatements:      float64(insertStatements.Load()),
			Postgres1:             float64(insertPostgres1.Load()),
			Postgres2:             float64(insertPostgres2.Load()),
		},
		Queries: QueryStats{
			Count:           float64(queryCount.Load()),
			TotalLatencySec: float64(qLat) / 1e6,
			FailedCount:     float64(queryFailed.Load()),
		},
	}
}

// Reporter holds state for the progress reporting goroutine and logs insert/query progress every interval.
type Reporter struct {
	Interval          time.Duration
	prevInserted      InsertedStats
	prevInsertStarted int64
	prevPostgres1     int64
	prevPostgres2     int64
	prevQueries       float64
	prevQueryLatency  float64
	prevFailed        float64
}

// NewReporter creates a Reporter with the given log interval. If interval <= 0, defaultInterval is used.
func NewReporter(interval time.Duration) *Reporter {
	if interval <= 0 {
		interval = defaultInterval
	}
	return &Reporter{Interval: interval}
}

// Run runs in the calling goroutine and logs insert/query progress every r.Interval.
// When doneCh is closed, it sends the final Snapshot on resultCh and closes resultCh. resultCh must be buffered (e.g. capacity 1).
func (r *Reporter) Run(doneCh <-chan struct{}, resultCh chan<- Snapshot) {
	ticker := time.NewTicker(r.Interval)
	defer ticker.Stop()

	for {
		select {
		case <-doneCh:
			resultCh <- loadSnapshot()
			close(resultCh)
			return
		case <-ticker.C:
			snap := loadSnapshot()
			total := snap.Inserted.Total
			originals := snap.Inserted.Originals
			duplicates := snap.Inserted.Duplicates
			totalInsertLatency := snap.Inserted.TotalInsertLatencySec
			insertStatements := snap.Inserted.InsertStatements
			q := snap.Queries.Count
			totalQueryLatency := snap.Queries.TotalLatencySec
			failed := snap.Queries.FailedCount
			curPostgres1 := insertPostgres1.Load()
			curPostgres2 := insertPostgres2.Load()

			curInsertStarted := insertStarted.Load()
			intervalInsertStarted := int(curInsertStarted - r.prevInsertStarted)
			r.prevInsertStarted = curInsertStarted

			intervalTotal := int(total - r.prevInserted.Total)
			intervalOriginals := int(originals - r.prevInserted.Originals)
			intervalDuplicates := int(duplicates - r.prevInserted.Duplicates)
			intervalLatency := totalInsertLatency - r.prevInserted.TotalInsertLatencySec
			intervalStatements := int(insertStatements - r.prevInserted.InsertStatements)
			r.prevInserted = InsertedStats{total, originals, duplicates, totalInsertLatency, insertStatements, float64(curPostgres1), float64(curPostgres2)}

			intervalAvgInsertMs := 0.0
			if intervalTotal > 0 {
				intervalAvgInsertMs = intervalLatency / float64(intervalTotal) * 1000
			}
			cumulativeAvgInsertMs := 0.0
			if total > 0 {
				cumulativeAvgInsertMs = totalInsertLatency / total * 1000
			}
			intervalPostgres1 := int(curPostgres1 - r.prevPostgres1)
			intervalPostgres2 := int(curPostgres2 - r.prevPostgres2)
			r.prevPostgres1 = curPostgres1
			r.prevPostgres2 = curPostgres2

			intervalQ := int(q - r.prevQueries)
			intervalQueryLatency := totalQueryLatency - r.prevQueryLatency
			intervalFailed := int(failed - r.prevFailed)
			r.prevQueries = q
			r.prevQueryLatency = totalQueryLatency
			r.prevFailed = failed
			avgLatencyMs := 0.0
			if q > 0 {
				avgLatencyMs = totalQueryLatency / q * 1000
			}
			intervalAvgMs := 0.0
			if intervalQ > 0 {
				intervalAvgMs = intervalQueryLatency / float64(intervalQ) * 1000
			}

			colW := 12
			log.Printf("%s---%s", _colorDim, _colorReset)
			log.Println(_colorYellow + "  Insert   " + padLeft("incoming", colW) + padLeft("completed", colW) + " " +
				padLeft("int_tot", colW) + padLeft("int_orig", colW) + padLeft("int_dup", colW) + padLeft("int_avg_ms", colW) + " " +
				padLeft("cum_tot", colW) + padLeft("cum_orig", colW) + padLeft("cum_dup", colW) + padLeft("cum_avg_ms", colW) + _colorReset)
			log.Printf("           %s%*d%s%s%*d%s %s%*d%s%s%*d%s%s%*d%s%s%*.*f%s %s%*d%s%s%*d%s%s%*d%s%s%*.*f%s",
				_colorCyan, colW, intervalInsertStarted, _colorReset,
				_colorCyan, colW, intervalStatements, _colorReset,
				_colorCyan, colW, intervalTotal, _colorReset,
				_colorCyan, colW, intervalOriginals, _colorReset,
				_colorCyan, colW, intervalDuplicates, _colorReset,
				_colorCyan, colW, 2, intervalAvgInsertMs, _colorReset,
				_colorCyan, colW, int(total), _colorReset,
				_colorCyan, colW, int(originals), _colorReset,
				_colorCyan, colW, int(duplicates), _colorReset,
				_colorCyan, colW, 2, cumulativeAvgInsertMs, _colorReset)
			log.Printf("  DB       postgres1: int %s%*d%s cum %s%*d%s   postgres2: int %s%*d%s cum %s%*d%s",
				_colorCyan, colW, intervalPostgres1, _colorReset,
				_colorCyan, colW, int(curPostgres1), _colorReset,
				_colorCyan, colW, intervalPostgres2, _colorReset,
				_colorCyan, colW, int(curPostgres2), _colorReset)
			log.Println(_colorYellow + "  Query    " + padLeft("int_queries", colW) + padLeft("int_failed", colW) + padLeft("int_avg_ms", colW) + " " +
				padLeft("cum_queries", colW) + padLeft("cum_failed", colW) + padLeft("cum_avg_ms", colW) + _colorReset)
			log.Printf("           %s%*d%s%s%*d%s%s%*.*f%s %s%*.0f%s%s%*.0f%s%s%*.*f%s",
				_colorCyan, colW, intervalQ, _colorReset,
				_colorCyan, colW, intervalFailed, _colorReset,
				_colorCyan, colW, 2, intervalAvgMs, _colorReset,
				_colorCyan, colW, q, _colorReset,
				_colorCyan, colW, failed, _colorReset,
				_colorCyan, colW, 2, avgLatencyMs, _colorReset)
		}
	}
}
